brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+        elif not self.GP_inten or inten_weight == 0:
+            self.U_, self.L_, self.nSNR_, self.sigma_, self.rho_,\
+                self.lGPspace_, self.bGP_ = self._fit_RSA_UV(
+                    design, X, scan_onsets=scan_onsets, coords=coords)
+        else:
+            self.U_, self.L_, self.nSNR_, self.sigma_, self.rho_, \
+                self.lGPspace_, self.bGP_, self.lGPinten_ = \
+                self._fit_RSA_UV(design, X, scan_onsets=scan_onsets,
+                                 coords=coords, inten=inten,
+                                 inten_weight=inten_weight)
+
+        if self.pad_DC:
+            self.U_ = self.U_[:-1, :-1]
+            self.L_ = self.L_[:-1, :self.rank]
+        sqrt_diag = np.diag(self.U_)**0.5
+        self.C_ = self.U_ / np.tile(sqrt_diag, [np.size(sqrt_diag), 1])\
+            / np.transpose(np.tile(sqrt_diag, [np.size(sqrt_diag), 1]))
+        return self
+
+    # The following 2 functions below generate templates used
+    # for constructing inverse of covariance matrix of AR(1) noise
+
+    def _D_gen(self, TR):
+        if TR > 0:
+            return np.diag(np.ones(TR - 1), -1) \
+                + np.diag(np.ones(TR - 1), 1)
+        else:
+            return np.empty([0, 0])
+
+    def _F_gen(self, TR):
+        if TR > 0:
+            F = np.eye(TR)
+            F[0, 0] = 0
+            F[TR - 1, TR - 1] = 0
+            return F
+        else:
+            return np.empty([0, 0])
+
+    def _prepare_data(self, design, Y, n_T, n_V, scan_onsets=None):
+        """Prepares different forms of products of design matrix and data,
+        or between themselves. These products are reused a lot during fitting.
+        So we pre-calculate them. Because of the fact that these are reused,
+        it is in principle possible to update the fitting as new data come in,
+        by just incrementally adding the products of new data and
+        their corresponding part of design matrix
+        """
+        if scan_onsets is None:
+            # assume that all data are acquired within the same scan.
+            D = np.diag(np.ones(n_T - 1), -1) + np.diag(np.ones(n_T - 1), 1)
+            F = np.eye(n_T)
+            F[0, 0] = 0
+            F[n_T - 1, n_T - 1] = 0
+        else:
+            # Each value in the scan_onsets tells the index at which
+            # a new scan starts. For example, if n_T = 500, and
+            # scan_onsets = [0,100,200,400], this means that the time points
+            # of 0-99 are from the first scan, 100-199 are from the second,
+            # 200-399 are from the third and 400-499 are from the fourth
+            run_TRs = np.diff(np.append(scan_onsets, n_T))
+            logger.debug('I infer that the number of volumes'
+                         ' in each scan are: {}'.format(run_TRs))
+
+            D_ele = list(map(self._D_gen, run_TRs))
+            F_ele = list(map(self._F_gen, run_TRs))
```

Comment:
Do you need to turn these iterators to lists or can you just iterate on them directly?